marker_api/src/ast/expr/ctor_expr.rs

```diff
@@ -0,0 +1,277 @@
+/// //      ^^^^^^^^^^^^^^^^^ A tuple constructor with two elements, represented
+/// //                        with field names, as above.
+/// let _ = TupleStruct { 0: 3, ..TupleStruct::default() };
+/// //      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+/// //      A tuple struct constructor with an optional base expression
+///
+/// let _ = Enum::A;
+/// //      ^^^^^^^ An enum variant constructor without any elements
+/// let _ = Enum::B(1);
+/// //      ^^^^^^^^^^ An enum variant constructor with two elements
+/// let _ = Enum::C { f1: 44, f2: 55 };
+/// //      ^^^^^^^^^^^^^^^^^^^^^^^^^^ An enum variant constructor with named fields
+/// ```
+#[repr(C)]
+#[derive(Debug)]
+pub struct CtorExpr<'ast> {
+    data: CommonExprData<'ast>,
+    path: AstQPath<'ast>,
+    fields: FfiSlice<'ast, CtorField<'ast>>,
+    base: FfiOption<ExprKind<'ast>>,
+}
+
+impl<'ast> CtorExpr<'ast> {
+    /// The path will point to the
+    pub fn path(&self) -> &AstQPath<'ast> {
+        &self.path
+    }
+
+    pub fn fields(&self) -> &'ast [CtorField<'ast>] {
+        self.fields.get()
+    }
+
+    pub fn base(&self) -> Option<ExprKind<'ast>> {
+        self.base.copy()
+    }
+}
+
+super::impl_expr_data!(
+    CtorExpr<'ast>,
+    Ctor,
+    fn precedence(&self) -> ExprPrecedence {
+        ExprPrecedence::Ctor
+    }
+);
+
+#[cfg(feature = "driver-api")]
+impl<'ast> CtorExpr<'ast> {
+    pub fn new(
+        data: CommonExprData<'ast>,
+        path: AstQPath<'ast>,
+        fields: &'ast [CtorField<'ast>],
+        base: Option<ExprKind<'ast>>,
+    ) -> Self {
+        Self {
+            data,
+            path,
+            fields: fields.into(),
+            base: base.into(),
+        }
+    }
+}
+
+/// A single field inside a [`CtorExpr`].
+#[repr(C)]
+#[derive(Debug)]
+pub struct CtorField<'ast> {
+    span: SpanId,
+    ident: Ident<'ast>,
+    expr: ExprKind<'ast>,
+}
+
+impl<'ast> CtorField<'ast> {
+    /// This returns the span of the entire field expression
+    pub fn span(&self) -> &Span<'ast> {
+        with_cx(self, |cx| cx.get_span(self.span))
+    }
+
+    /// The identifier of the field.
+    pub fn ident(&self) -> &Ident<'ast> {
+        &self.ident
+    }
+
+    pub fn expr(&self) -> ExprKind<'ast> {
+        self.expr
+    }
+}
+
+#[cfg(feature = "driver-api")]
+impl<'ast> CtorField<'ast> {
+    pub fn new(span: SpanId, ident: Ident<'ast>, expr: ExprKind<'ast>) -> Self {
+        Self { span, ident, expr }
+    }
+}
+
+/// A range expression, like these:
+///
+/// ```
+/// 1..9;
+/// 3..;
+/// ..5;
+/// ..;
+/// 0..=1;
+/// ```
+#[repr(C)]
+#[derive(Debug)]
+pub struct RangeExpr<'ast> {
+    data: CommonExprData<'ast>,
+    start: FfiOption<ExprKind<'ast>>,
+    end: FfiOption<ExprKind<'ast>>,
+    is_inclusive: bool,
+}
+
+impl<'ast> RangeExpr<'ast> {
+    pub fn start(&self) -> Option<ExprKind<'ast>> {
+        self.start.copy()
+    }
+
+    pub fn end(&self) -> Option<ExprKind<'ast>> {
+        self.start.copy()
+    }
```

Comment:
```suggestion
    pub fn end(&self) -> Option<ExprKind<'ast>> {
        self.end.copy()
    }
```
Copy-paste :sweat_smile:

Reply:
Maybe, I'm relying too much on your reviews, that I didn't double-check this. Thank you, this has been an enormous help as always!